File: app/src/main/java/com/gmail/fattazzo/formula1world/fragments/current/races/detail/pages/statistics/AbstractRaceLapTimeFragment.java
package com.gmail.fattazzo.formula1world.fragments.current.races.detail.pages.statistics;

import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.v4.app.Fragment;
import android.view.View;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.ViewFlipper;

import com.github.mikephil.charting.animation.Easing;
import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
import com.gmail.fattazzo.formula1world.R;
import com.gmail.fattazzo.formula1world.domain.F1Constructor;
import com.gmail.fattazzo.formula1world.domain.F1Driver;
import com.gmail.fattazzo.formula1world.domain.F1LapTime;
import com.gmail.fattazzo.formula1world.domain.F1Race;
import com.gmail.fattazzo.formula1world.domain.F1Result;
import com.gmail.fattazzo.formula1world.fragments.current.races.detail.pages.statistics.positions.DriverSpinnerModel;
import com.gmail.fattazzo.formula1world.fragments.current.races.detail.pages.statistics.positions.DriversSpinnerAdapter;
import com.gmail.fattazzo.formula1world.service.DataService;
import com.gmail.fattazzo.formula1world.utils.ThemeUtils;
import com.gmail.fattazzo.formula1world.view.chart.marker.F1MarkerLapTimeView;

import org.androidannotations.annotations.AfterInject;
import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Background;
import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.CheckedChange;
import org.androidannotations.annotations.Click;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.FragmentArg;
import org.androidannotations.annotations.UiThread;
import org.androidannotations.annotations.ViewById;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.collections4.map.HashedMap;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

/**
 * @author fattazzo
 *         <p/>
 *         date: 19/09/17
 */
@EFragment(R.layout.fragment_race_stat_chart_drivers)
public abstract class AbstractRaceLapTimeFragment extends Fragment {

    @FragmentArg
    protected F1Race race;

    @Bean
    protected DataService dataService;

    @Bean
    protected ThemeUtils themeUtils;

    @Bean
    protected DriversSpinnerAdapter spinnerAdapter;

    @ViewById
    protected Spinner drivers_spinner;

    @ViewById
    protected ConstraintLayout listHeader;

    @Bean
    protected LapTimeListAdapter adapter;

    @ViewById
    protected LineChart chart;
    @ViewById
    protected ListView listview;

    @ViewById
    protected ViewFlipper view_flipper;

    @ViewById
    protected ProgressBar refresh_progressBar;
    protected int textColor;
    protected float textSize;
    private DriverSpinnerModel allDriversSpinnerModel;

    @AfterInject
    protected void init() {
        textColor = themeUtils.getThemeTextColor(getContext());
        textSize = themeUtils.getThemeTextSize(getContext(), R.dimen.font_size_small);

        F1Driver f1Driver = new F1Driver();
        f1Driver.driverRef = "___";
        f1Driver.givenName = getString(R.string.all_drivers);
        f1Driver.familyName = "";
        F1Constructor f1Constructor = new F1Constructor();
        f1Constructor.constructorRef = "___";
        allDriversSpinnerModel = new DriverSpinnerModel(f1Driver, f1Constructor);
    }

    @AfterViews
    protected void initViews() {
        Set<DriverSpinnerModel> driversSpinner = new TreeSet<>(new Comparator<DriverSpinnerModel>() {
            @Override
            public int compare(DriverSpinnerModel o1, DriverSpinnerModel o2) {
                return o1.getDriver().getFullName().compareTo(o2.getDriver().getFullName());
            }
        });

        List<F1Result> results = dataService.loadRaceResult(race);
        if (dataService.hasLocalLapsData(race)) {
            driversSpinner.add(allDriversSpinnerModel);
        }
        for (F1Result result : results) {
            driversSpinner.add(new DriverSpinnerModel(result.driver, result.constructor));
        }

        spinnerAdapter.setDrivers(driversSpinner);
        drivers_spinner.setAdapter(spinnerAdapter);

        configureChart(chart);

        ((TextView) listHeader.findViewById(R.id.driverName)).setText(getString(R.string.drivers));
        ((TextView) listHeader.findViewById(R.id.lap)).setText(getString(R.string.lap));
        ((TextView) listHeader.findViewById(R.id.position)).setText(getString(R.string.stats_pos));
        ((TextView) listHeader.findViewById(R.id.time)).setText(getString(R.string.time2));
    }

    @CheckedChange
    protected void chartListviewSwitch(Switch chartListviewSwitch) {
        if (chartListviewSwitch.isChecked()) {
            view_flipper.showNext();
        } else {
            view_flipper.showPrevious();
        }
        view_flipper.invalidate();
    }

    @Click
    protected void add_button() {
        List<DriverSpinnerModel> driversToAdd = new ArrayList<>();
        if (drivers_spinner.getSelectedItem().equals(allDriversSpinnerModel)) {
            driversToAdd.addAll(((DriversSpinnerAdapter) drivers_spinner.getAdapter()).getItems());
        } else {
            driversToAdd.add((DriverSpinnerModel) drivers_spinner.getSelectedItem());
        }

        List<DriverSpinnerModel> driversForChart = new ArrayList<>();
        for (DriverSpinnerModel driverSelected : driversToAdd) {
            if (driverSelected.getDriver().driverRef.equals("___")) {
                // skip "all drivers" entry
                continue;
            }

            ILineDataSet dataSetDriver = chart.getLineData() != null ? chart.getLineData().getDataSetByLabel(driverSelected.getDriver().getFullName(), false) : null;
            if (dataSetDriver != null) {
                // dataset fro driver selected already added
                continue;
            }

            driversForChart.add(driverSelected);
        }
        loadDriverLapTimes(race, driversForChart);
    }

    @Click
    protected void remove_button() {
        DriverSpinnerModel driverSelected = (DriverSpinnerModel) drivers_spinner.getSelectedItem();

        try {
            chart.setMarker(null);

            if (driverSelected.getDriver().driverRef.equals("___")) {
                chart.clear();
                adapter.clearItems();
                adapter.notifyDataSetChanged();
            } else {
                ILineDataSet dataSetByLabel = chart.getLineData() != null ? chart.getLineData().getDataSetByLabel(driverSelected.getDriver().getFullName(), false) : null;
                if (dataSetByLabel != null) {
                    chart.getLineData().removeDataSet(dataSetByLabel);
                    chart.getLineData().notifyDataChanged();
                    chart.notifyDataSetChanged();

                    adapter.removeItems(driverSelected.getDriver());
                    adapter.notifyDataSetChanged();
                }
            }
        } finally {
            chart.animateX(1000, Easing.EasingOption.EaseInSine);

            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    if (chart != null) {
                        chart.setMarker(new F1MarkerLapTimeView(getContext()));
                    }
                }
            }, 2000);
        }
    }

    @Background
    protected void loadDriverLapTimes(F1Race race, List<DriverSpinnerModel> driversModel) {
        startAdd();

        Map<F1Driver, List<F1LapTime>> lapsMap = new HashedMap<>();
        for (DriverSpinnerModel model : driversModel) {
            List<F1LapTime> driverLapTimes = dataService.loadLaps(race, model.getDriver());
            if (CollectionUtils.isNotEmpty(driverLapTimes)) {
                lapsMap.put(model.getDriver(), driverLapTimes);
            }
        }

        updateChart(lapsMap);
        updateListView(lapsMap);
    }

    @UiThread
    protected void startAdd() {
        if (refresh_progressBar != null) {
            refresh_progressBar.setVisibility(View.VISIBLE);
        }
    }

    @UiThread
    protected void updateListView(Map<F1Driver, List<F1LapTime>> laps) {
        List<F1LapTime> lapsListView = new ArrayList<>();
        for (Map.Entry<F1Driver, List<F1LapTime>> driverEntry : laps.entrySet()) {
            for (F1LapTime lapTime : driverEntry.getValue()) {
                lapTime.driver = driverEntry.getKey();
                lapsListView.add(lapTime);
            }
        }

        sortLapTimesForList(lapsListView);

        adapter.setLapTimes(lapsListView);
        listview.setAdapter(adapter);
    }

    @UiThread
    protected void updateChart(Map<F1Driver, List<F1LapTime>> laps) {
        try {
            for (Map.Entry<F1Driver, List<F1LapTime>> driverEntry : laps.entrySet()) {
                LineDataSet dataSet = buildDataSet(driverEntry.getValue(), driverEntry.getKey());
                if (chart == null) {
                    return;
                }
                if (chart.getLineData() == null) {
                    chart.setData(new LineData(dataSet));
                } else {
                    chart.getLineData().addDataSet(dataSet);
                }
            }
            chart.notifyDataSetChanged();
            chart.animateX(1000, Easing.EasingOption.EaseInSine);
        } finally {
            if (refresh_progressBar != null) {
                refresh_progressBar.setVisibility(View.INVISIBLE);
            }
        }
    }

    protected abstract void configureChart(LineChart chart);

    protected abstract
    @NonNull
    LineDataSet buildDataSet(@NonNull List<F1LapTime> results, F1Driver driver);

    protected abstract void sortLapTimesForList(List<F1LapTime> lapsListView);
}

File: app/src/main/kotlin/com/gmail/fattazzo/formula1world/fragments/current/races/detail/pages/statistics/AbstractRaceLapTimeFragment.kt
package com.gmail.fattazzo.formula1world.fragments.current.races.detail.pages.statistics

import android.os.Handler
import android.support.constraint.ConstraintLayout
import android.support.v4.app.Fragment
import android.view.View
import android.widget.*
import com.github.mikephil.charting.animation.Easing
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.gmail.fattazzo.formula1world.R
import com.gmail.fattazzo.formula1world.domain.F1Constructor
import com.gmail.fattazzo.formula1world.domain.F1Driver
import com.gmail.fattazzo.formula1world.domain.F1LapTime
import com.gmail.fattazzo.formula1world.domain.F1Race
import com.gmail.fattazzo.formula1world.fragments.current.races.detail.pages.statistics.positions.DriverSpinnerModel
import com.gmail.fattazzo.formula1world.fragments.current.races.detail.pages.statistics.positions.DriversSpinnerAdapter
import com.gmail.fattazzo.formula1world.service.DataService
import com.gmail.fattazzo.formula1world.utils.ThemeUtils
import com.gmail.fattazzo.formula1world.view.chart.marker.F1MarkerLapTimeView
import org.androidannotations.annotations.*
import org.apache.commons.collections4.CollectionUtils
import org.apache.commons.collections4.map.HashedMap
import java.util.*

/**
 * @author fattazzo
 *
 *
 * date: 19/09/17
 */
@EFragment(R.layout.fragment_race_stat_chart_drivers)
abstract class AbstractRaceLapTimeFragment : Fragment() {

    @FragmentArg
    lateinit protected var race: F1Race

    @Bean
    lateinit protected var dataService: DataService

    @Bean
    lateinit protected var themeUtils: ThemeUtils

    @Bean
    lateinit protected var spinnerAdapter: DriversSpinnerAdapter

    @ViewById
    lateinit protected var drivers_spinner: Spinner

    @ViewById
    lateinit protected var listHeader: ConstraintLayout

    @Bean
    lateinit protected var adapter: LapTimeListAdapter

    @ViewById
    lateinit protected var chart: LineChart
    @ViewById
    lateinit protected var listview: ListView

    @ViewById
    lateinit protected var view_flipper: ViewFlipper

    @ViewById
    lateinit protected var refresh_progressBar: ProgressBar
    protected var textColor: Int = 0
    protected var textSize: Float = 0.toFloat()
    private var allDriversSpinnerModel: DriverSpinnerModel? = null

    @AfterInject
    protected fun init() {
        textColor = themeUtils!!.getThemeTextColor(context)
        textSize = themeUtils!!.getThemeTextSize(context, R.dimen.font_size_small)

        val f1Driver = F1Driver()
        f1Driver.driverRef = "___"
        f1Driver.givenName = getString(R.string.all_drivers)
        f1Driver.familyName = ""
        val f1Constructor = F1Constructor()
        f1Constructor.constructorRef = "___"
        allDriversSpinnerModel = DriverSpinnerModel(f1Driver, f1Constructor)
    }

    @AfterViews
    protected fun initViews() {
        val driversSpinner = TreeSet(Comparator<DriverSpinnerModel> { o1, o2 -> o1.driver!!.fullName.compareTo(o2.driver!!.fullName) })

        val results = dataService!!.loadRaceResult(race!!)
        if (dataService!!.hasLocalLapsData(race!!)) {
            driversSpinner.add(allDriversSpinnerModel)
        }
        for (result in results) {
            driversSpinner.add(DriverSpinnerModel(result.driver, result.constructor))
        }

        spinnerAdapter!!.setDrivers(driversSpinner)
        drivers_spinner!!.adapter = spinnerAdapter

        configureChart(chart)

        (listHeader!!.findViewById<TextView>(R.id.driverName)).text = getString(R.string.drivers)
        (listHeader!!.findViewById<TextView>(R.id.lap)).text = getString(R.string.lap)
        (listHeader!!.findViewById<TextView>(R.id.position)).text = getString(R.string.stats_pos)
        (listHeader!!.findViewById<TextView>(R.id.time)).text = getString(R.string.time2)
    }

    @CheckedChange
    protected fun chartListviewSwitch(chartListviewSwitch: Switch) {
        if (chartListviewSwitch.isChecked) {
            view_flipper!!.showNext()
        } else {
            view_flipper!!.showPrevious()
        }
        view_flipper!!.invalidate()
    }

    @Click
    protected fun add_button() {
        val driversToAdd = ArrayList<DriverSpinnerModel>()
        if (drivers_spinner!!.selectedItem == allDriversSpinnerModel) {
            driversToAdd.addAll((drivers_spinner!!.adapter as DriversSpinnerAdapter).items!!)
        } else {
            driversToAdd.add(drivers_spinner!!.selectedItem as DriverSpinnerModel)
        }

        val driversForChart = ArrayList<DriverSpinnerModel>()
        for (driverSelected in driversToAdd) {
            if (driverSelected.driver!!.driverRef == "___") {
                // skip "all drivers" entry
                continue
            }

            val dataSetDriver = if (chart!!.lineData != null) chart!!.lineData.getDataSetByLabel(driverSelected.driver!!.fullName, false) else null
            if (dataSetDriver != null) {
                // dataset fro driver selected already added
                continue
            }

            driversForChart.add(driverSelected)
        }
        loadDriverLapTimes(race, driversForChart)
    }

    @Click
    protected fun remove_button() {
        val driverSelected = drivers_spinner!!.selectedItem as DriverSpinnerModel

        try {
            chart!!.marker = null

            if (driverSelected.driver!!.driverRef == "___") {
                chart!!.clear()
                adapter!!.clearItems()
                adapter!!.notifyDataSetChanged()
            } else {
                val dataSetByLabel = if (chart!!.lineData != null) chart!!.lineData.getDataSetByLabel(driverSelected.driver!!.fullName, false) else null
                if (dataSetByLabel != null) {
                    chart!!.lineData.removeDataSet(dataSetByLabel)
                    chart!!.lineData.notifyDataChanged()
                    chart!!.notifyDataSetChanged()

                    adapter!!.removeItems(driverSelected.driver!!)
                    adapter!!.notifyDataSetChanged()
                }
            }
        } finally {
            chart!!.animateX(1000, Easing.EasingOption.EaseInSine)

            Handler().postDelayed({
                if (chart != null) {
                    chart!!.marker = F1MarkerLapTimeView(context)
                }
            }, 2000)
        }
    }

    @Background
    protected open fun loadDriverLapTimes(race: F1Race?, driversModel: List<DriverSpinnerModel>) {
        startAdd()

        val lapsMap = HashedMap<F1Driver, List<F1LapTime>>()
        for (model in driversModel) {
            val driverLapTimes = dataService!!.loadLaps(race!!, model.driver!!)
            if (CollectionUtils.isNotEmpty(driverLapTimes)) {
                lapsMap.put(model.driver, driverLapTimes)
            }
        }

        updateChart(lapsMap)
        updateListView(lapsMap)
    }

    @UiThread
    protected open fun startAdd() {
        if (refresh_progressBar != null) {
            refresh_progressBar!!.visibility = View.VISIBLE
        }
    }

    @UiThread
    protected open fun updateListView(laps: Map<F1Driver, List<F1LapTime>>) {
        val lapsListView = ArrayList<F1LapTime>()
        for ((key, value) in laps) {
            for (lapTime in value) {
                lapTime.driver = key
                lapsListView.add(lapTime)
            }
        }

        sortLapTimesForList(lapsListView)

        adapter!!.setLapTimes(lapsListView)
        listview!!.adapter = adapter
    }

    @UiThread
    protected open fun updateChart(laps: Map<F1Driver, List<F1LapTime>>) {
        try {
            for ((key, value) in laps) {
                val dataSet = buildDataSet(value, key)
                if (chart == null) {
                    return
                }
                if (chart!!.lineData == null) {
                    chart!!.data = LineData(dataSet)
                } else {
                    chart!!.lineData.addDataSet(dataSet)
                }
            }
            chart!!.notifyDataSetChanged()
            chart!!.animateX(1000, Easing.EasingOption.EaseInSine)
        } finally {
            if (refresh_progressBar != null) {
                refresh_progressBar!!.visibility = View.INVISIBLE
            }
        }
    }

    protected abstract fun configureChart(chart: LineChart?)

    protected abstract fun buildDataSet(results: List<F1LapTime>, driver: F1Driver): LineDataSet

    protected abstract fun sortLapTimesForList(lapsListView: List<F1LapTime>)
}
